src/exe/Main.hs

```diff
@@ -85,28 +76,64 @@ compiler = do
     putDebug "No input file.\n"
     putUsageInfo
     exitSuccess
+  -- Select frontend and backend
+  frontend <- selectFrontend
+  backend  <- selectBackend
   -- Initialize environment.
   loadPrelude
   loadQuickCheck
-  createCoqProject
+  specialAction backend
   -- Process input files.
-  modules  <- inOpts optInputFiles >>= mapM parseInputFile >>= sortInputModules
-  modules' <- mapM convertInputModule modules
-  mapM_ (uncurry outputCoqModule) modules'
+  modules' <-
+    inOpts optInputFiles
+    >>= mapM (parseInputFile $ parseFile frontend)
+    >>= sortInputModules
+    >>= mapM (liftConverter . runPipeline)
+    >>= mapM (convertInputModule $ convertModule backend)
+  mapM_ (uncurry (outputModule $ fileExtension backend)) modules'
+
+-------------------------------------------------------------------------------
+-- Front- and backend selection                                              --
+-------------------------------------------------------------------------------
+
+-- | Selects the correct frontend or throws an error if such a frontend does
+--   not exist.
+selectFrontend :: Application Frontend
+selectFrontend = do
+  frontendName <- inOpts optFrontend
+  case Map.lookup frontendName frontends of
+    Nothing -> do
+      putDebug
+        $  "Unrecognized frontend. Currently supported frontends are: "
+        ++ intercalate ", " (Map.keys frontends)
+      liftIO exitFailure
+    Just f -> return f
+
+-- | Selects the correct backend or throws an error if such a backend does
+--   not exist.
+selectBackend :: Application Backend
+selectBackend = do
+  backendName <- inOpts optBackend
+  case Map.lookup backendName backends of
+    Nothing -> do
+      putDebug
+        $  "Unrecognized backend. Currently supported backends are: "
+        ++ intercalate ", " (Map.keys backends)
+      liftIO exitFailure
```

Comment:
This is not really a debugging message but an actual message that should be reported using `reportFatal`.